xtractmime/__init__.py

```diff
@@ -1,45 +1,179 @@
 __version__ = "0.0.0"
 from typing import Optional, Set, Tuple, Union
+from xtractmime._patterns import _APACHE_TYPES, WHITESPACE_BYTES
+from xtractmime._utils import (
+    contains_binary,
+    get_archive_mime,
+    get_audio_video_mime,
+    get_extra_mime,
+    get_font_mime,
+    get_image_mime,
+    get_text_mime,
+)
 
-_APACHE_TYPES = [
-    b"text/plain",
-    b"text/plain; charset=ISO-8859-1",
-    b"text/plain; charset=iso-8859-1",
-    b"text/plain; charset=UTF-8",
-]
-WHITESPACE_BYTES = {b"\t", b"\r", b"\x0c", b"\n", b" "}
 
+def _find_unknown_mimetype(
+    input_bytes: bytes,
+    sniff_scriptable: bool,
+    extra_types: Optional[Tuple[Tuple[bytes, bytes, Union[Set[bytes], None], bytes], ...]],
+) -> Optional[bytes]:
+    if sniff_scriptable:
+        matched_type = get_text_mime(input_bytes)
+        if matched_type:
+            return matched_type
 
-def _is_match_mime_pattern(
-    input_bytes: bytes, byte_pattern: bytes, pattern_mask: bytes, lstrip: Set[bytes] = None
-) -> bool:
+    matched_type = get_extra_mime(input_bytes, extra_types=extra_types)
+    if matched_type:
+        return matched_type
+
+    matched_type = get_image_mime(input_bytes)
+    if matched_type:
+        return matched_type
+
+    matched_type = get_audio_video_mime(input_bytes)
+    if matched_type:
+        return matched_type
+
+    matched_type = get_archive_mime(input_bytes)
+    if matched_type:
+        return matched_type
+
+    if not contains_binary(input_bytes):
+        return b"text/plain"
+
+    return b"application/octet-stream"
+
+
+def _sniff_mislabled_binary(input_bytes: bytes) -> Optional[bytes]:
+
+    if input_bytes[:2] in (b"\xfe\xff", b"\xff\xfe") or input_bytes[:3] == b"\xef\xbb\xbf":
+        return b"text/plain"
+
+    if not contains_binary(input_bytes):
+        return b"text/plain"
+
+    return b"application/octet-stream"
+
+
+def _sniff_mislabled_feed(input_bytes: bytes, supplied_type: bytes) -> Optional[bytes]:
     input_size = len(input_bytes)
-    pattern_size = len(byte_pattern)
-    mask_size = len(pattern_mask)
+    index = 0
+
+    if input_bytes[:3] == b"\xef\xbb\xbf":
+        index += 3
+
+    while index < input_size:
+        while True:
+            if not input_bytes[index : index + 1]:
+                return supplied_type
```

Comment:
I believe you can reach this line (test coverage) with a file that contains a single space as content (i.e. `b' '`).

Reply:
Thanks, it's working. Got 100% coverage as of now 😅